pandas/_config/localization.py

```diff
@@ -45,7 +45,9 @@ def set_locale(
         locale.setlocale(lc_var, new_locale)
         normalized_locale = locale.getlocale()
         if all(x is not None for x in normalized_locale):
-            yield ".".join(normalized_locale)
+            # error: Argument 1 to "join" of "str" has incompatible type
+            # "Tuple[Optional[str], Optional[str]]"; expected "Iterable[str]"
+            yield ".".join(normalized_locale)  # type: ignore[arg-type]
```

Comment:
mypy doesn't seem to understand `all(...)` to narrow the type.

Reply:
this works, but let's not do this yet! (until we can use it elsewhere)

```
        def all_not_none(arg: Iterable[str | None]) -> TypeGuard[Iterable[str]]:
            return all(x is not None for x in arg)

        if all_not_none(normalized_locale):
```

Reply:
looking through the code there are a couple of other places this pattern is used that do not use `all_not_none` from `pandas/core/common.py`

maybe change this one here, then in the future we only need the TypeGuard in one place.

```
    try:
        import pandas.core.common as com

        locale.setlocale(lc_var, new_locale)
        normalized_locale = locale.getlocale()
        if com.all_not_none(*normalized_locale):
```

note need to prevent circular import.

Reply:
Have there been discussions about using `typing_extensions`? Then we could already use TypeGuard.

Reply:
no we don't use typing_extensions but we can import from typing in pandas._typing with a guard so that it resolves to Any in older python versions.

Reply:
> Have there been discussions about using `typing_extensions`?

As long as `typing_extensions` is not a run-time dependency, I would not be adverse to it being added as a dev dependency.